src/c2.c

```diff
@@ -610,24 +610,30 @@ static int c2_parse_target(const char *pattern, int offset, c2_ptr_t *presult) {
 
 	// Parse index
 	if ('[' == pattern[offset]) {
+		if (pleaf->predef != C2_L_PUNDEFINED) {
+			c2_error("Predefined targets can't have index.");
```

Comment:
`wm_state` is going to be a predefined target, and it can have index.

Reply:
`wm_state` doesn't currently support indexing. That doesn't mean I couldn't add the support for indices. I'm not sure how useful indexing would be, though.

Reply:
I think it might be good for consistency. Because wm_state is indeed an array behind the sceen.

Reply:
So we would inherit the same behaviour (no index: first, index `*`: all) as with the custom properties? I can change and document that in #549.

Reply:
Yes, I think that would be good. Do you have a feeling on this?

Reply:
The initial idea was to make checks against specific values of `_NET_WM_STATE` easier, especially the hidden and sticky states. In contrast to adding a new predefined target for every of the possibly useful states (similar to `fullscreen` but only a light wrapper over the existence of said window property), I opted to add a new target analogous to `window_type` to keep this more consice. If you think the other approach would be more consistent I am open to discuss that in more detail in #549 (especially regarding the difference between `wm_state = 'fullscreen'` and `fullscreen`).
Regardless, all functionality regarding a new target for `wm_state` is already possible with this PR and direct querying of the `NET_WM_STATE` atom. If we opt to not add the new target/targets caching, of wm_state can be simplified to just caching if fullscreen is part of that atom. These targets would just be a convenience for the user and save us from querying the X server with every rule evaluation.

Reply:
OK, IIUC, caching of `wm_state` is to speed up the case where there are multiple `wm_state` matches in the exclusion rules. This does sound like a good thing to do, so :+1:

The other thing is whether we want `wm_state = 'fullscreen'` or `wm_state[*] = 'fullscreen'`, which I am not really sure about. I think the latter is more consistent, but would like your thoughts on this as well.

Reply:
I initially went for `wm_state = 'fullscreen'` since I don't really see any benefit from allowing array indexing:
1) Why would I need to differentiate between the `FULLSCREEN` atom being first or last (the spec does not indicate any priority indicated by the index).
2) The atom is only cached as a bitmask, loosing any order the array might have had.
Therefore, changing `wm_state` to use indices would be misleading in my opinion suggesting some order that isn't reflected by the window.

Furthermore, the usecase of the `wm_state` predicate would be a more readable (and cached) rule-checking against "does _NET_WM_STATE contain the hidden, sticky, maximized_vert, … atom" — i.e. a "shortcut" for `_NET_WM_STATE@[*]:32a = "_NET_WM_STATE_HIDDEN", etc.

To be honest, I am more concerned about users confusing `wm_state = 'fullscreen'` with the `fullscreen`-predicate (the latter additionally checks window dimensions to match screen dimensions), and `wm_state = 'focused'` with the `focused`-predicate, which is definitely not the same (the latter is our internal state influenced ,e.g., by `focus-exclude` while the former is just a flag set by the window manager). This could be countered by keeping the `fullscreen` and `focused` predicates and introducing new ones at least for `hidden`, `sticky`, and maybe `maximized_(vert|horz)` that just check if the corresponding bit is set in the cached `wm_state`.

I personally prefer the current approach of bundling them into a single `wm_state` predicate but can see how separate predicates would make for an easier rule-evaluation (just flag checking instead of string matching over an array).

Reply:
> To be honest, I am more concerned about users confusing ...

Good point, that plus `wm_state` being just a shortcut without other added values, makes me wonder if we should just ask the user to use `_NET_WM_STATE@[*]:32a = "_NET_WM_STATE_HIDDEN"`. User won't get wrong expectations from that rule.

Reply:
And for the caching aspect, I think we could add generic caching for all properties.

Reply:
> … `wm_state` being just a shortcut without other added values …

Yeah, this was just meant as a compromise to make these rules "easier to read" (see https://github.com/yshui/picom/issues/512#issuecomment-715445951). I am fine with leaving those changes out.

> I think we could add generic caching for all properties.

Let me see if this can be nicely done for arbitrary atom/string properties (and atom arrays) without adding too much overhead.

Reply:
Caching can be done on-demand when properties are used, eviction can be done when we get property notify.

Reply:
I'll take over #549 to implement general caching instead of the `wm_state` target.